server/core/src/finalize_utilities.cpp

```diff
@@ -223,25 +223,18 @@ namespace irods
 
     auto duplicate_l1_descriptor(const l1desc& _src) -> l1desc
     {
-        l1desc dest{};
-        std::memcpy(&dest, &_src, sizeof(l1desc));
+        l1desc dest;
 
-        if (_src.dataObjInp) {
-            DataObjInp* doi = static_cast<DataObjInp*>(std::malloc(sizeof(DataObjInp)));
-            replDataObjInp(_src.dataObjInp, doi);
-            dest.dataObjInp = doi;
-        }
+        init_l1desc(dest);
+        copy_l1desc(dest, _src);
 
-        if (_src.dataObjInfo) {
-            auto [obj, obj_lm] = irods::experimental::replica::duplicate_replica(*_src.dataObjInfo);
-            dest.dataObjInfo = obj_lm.release();
+        // Keep the first DataObjInfo in the linked list and free the rest.
+        if (dest.dataObjInfo) {
+            auto* rest = dest.dataObjInfo->next;
+            freeAllDataObjInfo(rest);
+            dest.dataObjInfo->next = nullptr;
         }
 
-        // TODO: need duplication logic
-        dest.remoteZoneHost = nullptr;
-        dest.otherDataObjInfo = nullptr;
-        dest.replDataObjInfo = nullptr;
-
```

Comment:
@alanking I changed the code to do a deep copy of the source l1desc object. Alternatively, I could copy construct the source l1desc object and keep the rest of the code as it was originally. See #6792 for the alternative solution.

Using the copy constructor seems better in the sense that it is likely faster than using `copy_l1desc()`.

Any thoughts? Primarily in regard to correctness.

_I'm going to rename `copy_l1desc()` to `deep_copy_l1desc()` or something that includes `deep` in the name._

Reply:
I think that the copy constructor in #6792 is fine/bettter, as you said. It will have the same effect as the existing code without the UB risk.

The deep copy would be super-extra-safe since the source could go away and leave us with dangling pointers in the destination L1 descriptor. That apparently hasn't been a problem for anything so far since we weren't making a deep copy before.

Reply:
Same here. Both implementations would likely work. And the original code seems to only care about the first `DataObjInfo`.

What makes you think we can dangle? The original implementation appears to allocate and copy the data it cares about from the source (i.e. L230 and L236 in the original impl).

Reply:
Well, maybe the concern is unfounded as #6754 deprecates the unused pointers.

Reply:
Restored some of the implementation to its original form due to the rebase and this discussion.